src/numerical/spline.rs

```diff
@@ -36,71 +39,188 @@ use util::non_macro::*;
+
+        // Pre-calculations
+        let mut h = vec![0f64; n];
+        let mut b = vec![0f64; n];
+        let mut v = vec![0f64; n];
+        let mut u = vec![0f64; n];
+        for i in 0..n {
+            if i == 0 {
+                h[i] = node_x[i + 1] - node_x[i];
+                b[i] = (node_y[i + 1] - node_y[i]) / h[i];
+            } else {
+                h[i] = node_x[i + 1] - node_x[i];
+                b[i] = (node_y[i + 1] - node_y[i]) / h[i];
+                v[i] = 2. * (h[i] + h[i - 1]);
+                u[i] = 6. * (b[i] - b[i - 1]);
+            }
+        }
+
+        // Tri-diagonal matrix
+        let mut m = matrix(vec![0f64; (n - 1) * (n - 1)], n - 1, n - 1, Col);
+        for i in 0..n - 2 {
+            m[(i, i)] = v[i + 1];
+            m[(i + 1, i)] = h[i + 1];
+            m[(i, i + 1)] = h[i + 1];
+        }
+        m[(n - 2, n - 2)] = v[n - 1];
+
+        // Calculate z
+        let z_inner = m.inv().unwrap() * Vec::from(&u[1..]).to_matrix();
+        let mut z = vec![0f64];
+        z.extend(&z_inner.data);
+        z.push(0f64);
+
+        // Declare empty spline
+        let mut s: Vec<Polynomial> = Vec::new();
+
+        // Main process
+        for i in 0..n {
+            // Memoization
+            let t_i = node_x[i];
+            let t_i1 = node_x[i + 1];
+            let z_i = z[i];
+            let z_i1 = z[i + 1];
+            let h_i = h[i];
+            let y_i = node_y[i];
+            let y_i1 = node_y[i + 1];
+            let temp1 = poly(vec![1f64, -t_i]);
+            let temp2 = poly(vec![1f64, -t_i1]);
+
+            let term1 = temp1.powi(3) * (z_i1 / (6f64 * h_i));
+            let term2 = temp2.powi(3) * (-z_i / (6f64 * h_i));
+            let term3 = temp1 * (y_i1 / h_i - z_i1 * h_i / 6.);
+            let term4 = temp2 * (-y_i / h_i + h_i * z_i / 6.0);
+
+            s.push(term1 + term2 + term3 + term4);
+        }
+        return s;
+    }
+
+    /// Evaluate cubic spline with value
+    ///
+    /// # Examples
+    /// ```
+    /// extern crate peroxide;
+    /// use peroxide::*;
+    ///
+    /// let x = c!(0.9, 1.3, 1.9, 2.1);
+    /// let y = c!(1.3, 1.5, 1.85, 2.1);
+    ///
+    /// let s = CubicSpline::from_x_and_y(x, y);
+    ///
+    /// s.eval(2.0);
+    /// ```
+    pub fn eval<T>(&self, x: T) -> f64
+    where
+        T: std::convert::Into<f64> + Copy,
+    {
+        let x = x.into();
+
+        let index = match self.polynomials.binary_search_by(|(range, _)| {
+            if range.contains(&x) {
+                core::cmp::Ordering::Equal
+            } else if x < range.start {
+                core::cmp::Ordering::Greater
+            } else {
+                core::cmp::Ordering::Less
+            }
+        }) {
+            Ok(index) => index,
+            Err(index) => max(0, min(index, self.polynomials.len() - 1)),
+        };
+
+        self.polynomials[index].1.eval(x)
+    }
+
+    pub fn extend_with_x_and_y(&mut self, node_x: Vector, node_y: Vector) {
```

Comment:
Here I'm also not so happy.

Reply:
Also, it could be `extend_with_nodes`. What is your opinion about it?